### ISSUE TITLE
Route::localizedUrl($localeKey) with L7 new route model route key binding

### ISSUE THREAD

**Issue by Okipa** [codezero-be/laravel-localized-routes#26]
Hi @ivanvermeyen ✋ 

I just switched my base app on L7 and I am currently switching to the new route model key binding (`page/{page:slug}`).

I am wondering if there would be a way to avoid to override the Model route key in the `getRouteKey()` method as it was required before. Indeed, I would like to get my pages from the slugs in the front and from the ids in the backoffice.

But currently, the `Route::localizedUrl($localeKey)` seems not be aware of this new system of binding and renders the localized URL with the route key defined in the `getRouteKey()` method without taking care of an eventual declared field (could be the field when declared and the result of this method when nothing is declare).

I don't if I am very clear, everything become such complex when we speak about routes 😄

**Comment by ivanvermeyen** [comment]
Hello,

I didn't find anything to confirm this yet, so I should try this out to be sure... but I suspect that `page/{page:slug}` only works outside in, so you trigger the URL and Laravel uses the slug. This allows you to use the slug on 1 page and the ID on another...

I think when using the `route()` helper, we will still need to rely on the `getRouteKey` method or pass it the actual slug manually. [This package also uses the `getRouteKey` method](https://github.com/codezero-be/laravel-localized-routes/blob/af7d62ed5d295aa86d80f5320d3d4098c4d4c832/src/LocalizedUrlGenerator.php#L404).

Another problem I suspect is that `page:slug` won't fetch the localized version of the slug when you use, for example, the `spatie/laravel-translatable` package. It should query something like `slug->en` (unless Laravel looks for a slug attribute on the model, but I doubt that).

I will dig a bit deeper and test your use case in more detail.

**Comment by Okipa** [comment]
Sure, we will have to handle this in the `->resolveRouteBinding($value, $field = null)` method, which can easily be done with the new transmitted `$field` param.

The problem remains in the following defined method in the model, which currently is the only way to get translated routes with the localized `url` returned by the `Route::localizedUrl($localeKey)` macro: 

```php
public function getRouteKey(): string
    {
        return $this->getTranslation('url', app()->getLocale());
    }
```

I wonder if there would be a way to return the localized URL from the macro, using the new `$field` param if it is defined, and the id if it not defined.

**Comment by trippo** [comment]
Are there news on this? it's impossible to convert automatically route like Route::get('{group:slug}' or Route::get('{group:slug}/{product:slug}' without set the getRouteKey function, but this is an error because if I must use also group::other_parameter i can't

**Comment by trippo** [comment]
If you implement resolveRouteBinding you can use:
` route(Route::current()->getName(), Route::current()->parameters(), true, $localeKey)`
instead of
`Route::localizedUrl($localeKey)`
and work fine

**Comment by ivanvermeyen** [comment]
@trippo your solution requires named routes, but it looks like a great workaround to use the custom route keys.
I still need to get some quality time with time package and setup a testing database to test these features (and how it all behaves).

**Comment by trippo** [comment]
First thanks for your time... I replace mcamara/laravel-localization package with your today...
The problem with the Route::localizedUrl($localeKey) function is it does not consider the [route model binding with customized key](https://laravel.com/docs/8.x/routing#customizing-the-default-key-name)  so if you have 
`Route::get('{group:slug}'` or  `Route::get('{group:slug}/{product:slug}'` routes the function return /en/1/4 instead /en/group-en-slug/product-en-slug because in LocalizedUrlGenerator the parameters lose the attribute... if you force getRouteKey on the model will work but loose the possibility to have other attributes like {product:code} in other routes...

**Comment by ivanvermeyen** [comment]
Hello,

I finally took some time to figure out how to get `Route::localizedUrl()` to play nicely with the custom route keys and I'm happy to say [that it's working](https://github.com/codezero-be/laravel-localized-routes/commit/5f52765c463439e524048cecadf338801bd4fa6d).

Just waiting for an issue to be resolved with Codacy, because it's blocking my test runs on Gihub Actions.

As soon as the test officially pass, I will tag a new release. 👍

**Comment by ivanvermeyen** [comment]
Tagged and released in 2.4.2 🥳

**Comment by Okipa** [comment]
@ivanvermeyen Nice ! Thanks !
